### PR TITLE
feat(#17): register_user, Storage redesigned to struct

### LINKED ISSUE #17: internal endpoint for user account creation
On the public GitHub API [there is no ability](https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28) to create GitHub account, however we should provide this ability in our API. I suggest to create an endpoint `POST /users` with body like this: `username` + `password` that will create a user account.

### PR DESCRIPTION
**Pull request by h1alexbel**
@l3r8yJ take a look, please

closes #17

---

## PR-Codex overview
**Focus:** Add user registration functionality and improve XML storage handling.

### Detailed summary
- Added `register_user` module for user registration.
- Improved XML storage initialization and handling.
- Updated dependencies and added `hamcrest`.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: server/src/objects/user.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+mod tests {
+    use anyhow::Result;
+
+    use crate::objects::user::User;
+
+    #[test]
+    fn returns_username() -> Result<()> {
+        let expected = "jeff";
+        let jeff = User::new(String::from(expected));
+        assert_eq!(
```
Comment (l3r8yJ) [inline server/src/objects/user.rs:63]: @h1alexbel might be we should add `hamcrest` to project?

File: server/src/xml/storage.rs [hunk 2]
```diff
@@ -19,37 +22,66 @@
...
+        let mut file = match File::create(location) {
+            Ok(file) => file,
+            Err(err) => {
+                panic!("fakehub storage failed to initialize in '{location}': {err}");
+            }
+        };
+        if let Err(err) = file.write_all(
+            "<root>\
+             <github><users/></github>\
+            </root>"
```
Comment (l3r8yJ) [inline server/src/xml/storage.rs:47]: @h1alexbel i think we can move it to some constant

**Review by l3r8yJ — CHANGES_REQUESTED** [review]
@h1alexbel take a look at my comments below, please

File: server/src/objects/user.rs (hunk 1 above)
Comment (h1alexbel) [reply server/src/objects/user.rs:63]: @l3r8yJ good idea. Let's try it

**Comment by h1alexbel** [comment]
@l3r8yJ take a look again, please

File: server/src/lib.rs [hunk 3]
```diff
@@ -22,16 +22,21 @@
...
+use crate::routes::register_user::register_user;
+use crate::xml::storage::Storage;
 
+mod objects;
 pub mod report;
 mod routes;
 mod xml;
+#[allow(unused_imports)]
+#[macro_use]
+extern crate hamcrest;
```
Comment (l3r8yJ) [inline server/src/lib.rs:39]: better to move it to `mod tests`

File: server/src/objects/user.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+// @todo #17:40min Apply XMLed user to the <users/> node in storage.
+//  We should apply XMLed user to the <users> XML node in storage. First we
+//  need to check that user with provided name does not exist, and only then
+//  apply it to the storage. Keep in mind that application function in the
+//  storage should be thread-safe (as well as #xml function). Don't forget to
+//  create unit tests that prove that.
+impl User {
+    pub async fn save(self) -> Result<()> {
+        info!("registering user @{}", self.username);
+        let xml = to_string(&self).unwrap();
```
Comment (l3r8yJ) [inline server/src/objects/user.rs:47]: let's avoid such `#unwrap` calls, can you set clippy or create puzzle for it?

File: server/src/xml/storage.rs [hunk 5]
```diff
@@ -19,42 +22,78 @@
...
-        Ok(file) => {
-            info!("'{location}' initialized");
-            file
+#[derive(Default)]
+#[allow(dead_code)]
+pub struct Storage {
+    pub(crate) path: String,
+}
+
+const INIT_XML: &str = "<root>\
```
Comment (l3r8yJ) [inline server/src/xml/storage.rs:34]: better to use multiline string

**Review by l3r8yJ — CHANGES_REQUESTED** [review]
@h1alexbel take a look, please

File: server/src/lib.rs (hunk 3 above)
Comment (h1alexbel) [reply server/src/lib.rs:39]: @l3r8yJ for some reason I can't move it to the tests, it's only allowed in `lib.rs` or `main.rs`

File: server/src/objects/user.rs (hunk 4 above)
Comment (h1alexbel) [reply server/src/objects/user.rs:47]: @l3r8yJ I will create one

**Comment by h1alexbel** [comment]
@l3r8yJ updated

**Pull request merged**